### DISCUSSION TITLE
float vector 精度问题

### DISCUSSION THREAD

**Discussion by Ice-YcY** [milvus-io/milvus#32737] [Q&A and General discussion]
2.3.x版本只支持floatvector
java sdk校验必须是float类型
但是存进vdb后发现是小数点后有19位

实际计算的逻辑是什么？

**Comment by xiaofan-luan** [comment]
float小数点后不会有19位，一半只有6-7位数字

**Reply by Ice-YcY** [reply]
我刚贴了图，实际在attu看是19位

**Reply by xiaofan-luan** [reply]
19 位，是实际的现实，float超过6-7位就丢失精度了。您可以研究一下float的格式就明白了

**Comment by Ice-YcY** [comment]
![image](https://github.com/milvus-io/milvus/assets/85332705/ffe77831-fd73-46fa-ad27-30eb73595a69)

**Comment by Ice-YcY** [comment]
![image](https://github.com/milvus-io/milvus/assets/85332705/d905b4d3-6340-4d8b-ab90-d4fecda324b0)

**Comment by Ice-YcY** [comment]
SDK校验：
![image](https://github.com/milvus-io/milvus/assets/85332705/acfb1f80-9bcf-4fcf-b6db-7304c9518a42)

**Comment by yhmo** [comment]
这是因为Attu使用的是node sdk。 javascript里，浮点型的数值是用64位表示的，虽然milvus返回的数据是float32，但attu显示出来就成了64位的显示。实际上小数点6，7位之后的数字你都可以忽略掉。

**Reply by Ice-YcY** [reply]
![image](https://github.com/milvus-io/milvus/assets/85332705/136f5777-4897-4493-9b0c-1ac718a98361)
1.attu实际小数点后位数不固定
2.我们embedding model会小数点后18位，如果只用7，8位的话，精度会有丢失

所以我们是得需要升级milvus到2.4.x,使用float16或者bfloat vector吗？

**Reply by yhmo** [reply]
向量检索不需要这么高的位数，向量检索的英文简称是什么，是ANN search(Approximate Nearest Neighbor)。本身就是一个“近似检索”，32位精度已经足够。使用64位不但存储增加一倍，内存增加一倍，计算的耗时也增加很多，所以Milvus内部对向量都是使用32位处理。

bfloat16是精度更低的表示，为了追求更高的运算性能和更小的存储空间。所以bfloat16根本不会提高你看到的精度。
float16目前在milvus是以float32的方式在计算，实际上精度也是float32的精度。

打个比方，我们比较两个人脸的相似度，看两个人像不像，难道你会去比较他们两人的头发数量是否相等么？

**Reply by Ice-YcY** [reply]
你说的32位精度是最新的2.4.0支持的Float16吗？

**Reply by Ice-YcY** [reply]
还有你说的向量检索不需要这么高的位数，这个不是很理解，
这个不像是头发的比较吧。
更高位的向量不是代表更准确的维度吗？
向量检索的近似性不是两个向量间的比较吗？ 如果可以使用高位向量的话，不是更可以知道是否近似吗？例如高清和马赛克的比较。
还是说向量检索7，8位的位数就是最优的了？ 多了也没有提升？

**Reply by xiaofan-luan** [reply]
milvus一直支持32 bit精度的float，但是32位精度对应在数据层面，只有6-7位数字。

**Reply by xiaofan-luan** [reply]
相似度的层面，一般模型层面上，有2-3位小数就比较精确了

**Reply by yhmo** [reply]
32位指的是float32，精确到小数点后6，7位。简而言之，这个精度对于“相似搜索”已经足够，没必要上float64。上float64意味着双倍的空间成本以及时间成本，没人愿意为了那所谓的一点点精度提高而花这种冤枉钱。

**Comment by AriesJin** [comment]
my milvus version is 2.4, using the py interface, why does query show 6 digits, but the search result is 15 digits? The internal implementation is different?

**Reply by yhmo** [reply]
Basically yes, the implimentation is different.

For query(), each float value of vectors is converted to numpy.float32 to show 6 digits.
https://github.com/milvus-io/pymilvus/blob/46f12b21c80efdfc718bc2ef86d867b710a06f1c/pymilvus/client/entity_helper.py#L716

But converting a python float value to numpy.float32 is time-consuming work, especially for high-dimension vectors. In earlier versions, we convert the values one by one, the performance is 10x slower than no-conversion. In recent, we have improved the performance by converting values in a batch by numpy.array(), but is still 4x slower than no-conversion.

Other SDKs such as Java sdk and Go sdk have no such conversion logic, so their query performance is much faster than Python SDK.

The search() interface doesn't have this conversion logic, because the performance of search() is really important to users. So, you will see 15-digit values returned by search().
